mkFirefox: add policy to accept added extensions

Mikilio opened the pull request:
This commit adds extensions added via
profiles.<name>.extensions.packages to the policy. In consequence, the extensions are added without complaint and don't need to be enabled manually.

### Description

While switching browser, I noticed that the current method of installing extensions, does not enable them. This is because the only firefox is allowed to install extensions, and it only provides one way to do so automatically.
As such, I have improved the method of installing extensions by using the intended way by mozilla to do this, without sacrificing reproducibility. This is achieved by "downloading" extensions from the nix store.

### NOTE:
- The extensionPath uuid constant became obsolete through this because extensions don't need to be bundled anymore.
This involves @rycee to change his convenient output derivations for firefox modules.
- This change relies on the fact that firefox modules from @rycee NUR repository expose a addonId. This needs to be documented somewhere
- `nix fmt .` made some unnecessary changes.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -p treefmt nixfmt deadnix keep-sorted --run treefmt`.

- [x] Code tested through `nix run .#tests -- test-all` or
      `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef). (not applicable)

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```
  Maintainers and Involved:
  @chayleaf
  @onny
  @brckd 
  @kira-bruneau

JuneStepp commented:
FYI, I've tried this before and ran into some issues. I can't remember for sure, but it may have been to do with extension updates. Just take this as a reminder to test well.

Mikilio commented:
I have forgotten about this PR after a while. To wrap this up, I'd like to confirm with @khaneliman and @IanHollow the following actions.

- move my may to add extensions from the profiles section to something higher.
- leave the current way to add extensions intact to not break compatibility and not force the requirement of having a addinId.

Mikilio committed:
mkFirefox: add policy to accept added extensions

This commit adds extensions added via
profiles.<name>.globalExtensions to the policy.
In consequence the extensions are added without complaint and don't need
to be enabled manually.

Co-authored-by: khaneliman <khaneliman@users.noreply.github.com>

khaneliman commented on modules/programs/firefox/mkFirefoxModule.nix:
```diff
@@ -1132,17 +1180,39 @@ in
+                    installation_mode = "force_installed";
+                    install_url = "file://${package.outPath}/share/mozilla/${extensionPath}/${package.addonId}.xpi";
+                  }
+                  // settings
```
This drops `settings`: `// settings` merges into the `nameValuePair` wrapper, not the policy value. Wrap the value instead: `lib.nameValuePair package.addonId ({ ... } // settings)`.

khaneliman commented on modules/programs/firefox/mkFirefoxModule.nix:
```diff
@@ -1132,17 +1180,39 @@ in
+                  }
+                  // settings
+                else
+                  { }
```
This branch returns `{}` for packages without `addonId`, but `listToAttrs` expects each mapped item to have `name` and `value`, so evaluation fails with `attribute 'name' missing`. Either reject these packages with a clear assertion/type error or handle them via a fallback path.

Mikilio commented:
@khaneliman I made some pretty embarrassing mistakes. I apologize for the extra review burden. This should fix those points mentioned.

khaneliman commented on modules/programs/firefox/mkFirefoxModule.nix:
```diff
@@ -1132,17 +1186,43 @@ in
+                map (
+                  elem:
+                  let
+                    package = elem.package or elem;
+                    settings = elem.settings or { };
+                  in
+                  if package ? addonId then
+                    lib.nameValuePair package.addonId (
+                      {
+                        installation_mode = "force_installed";
+                        install_url = "file://${package.outPath}/share/mozilla/${extensionPath}/${package.addonId}.xpi";
+                      }
+                      // settings
+                    )
+                  else
+                    { }
+                ) cfg.globalExtensions
+              )
+            )
```
This still silently drops packages without `addonId`, which conflicts with the option description saying `addonId` is required. You can probably add an assertion like this elsewhere to make sure all extensions are valid:

```nix
{
  assertion = builtins.all (
    elem:
    let
      package = elem.package or elem;
    in
    package ? addonId
  ) cfg.globalExtensions;
  message = "${moduleName}.globalExtensions requires each package to expose addonId in passthru.";
}
```

Then this highlighted block can stay direct:

```suggestion
lib.listToAttrs (
  map (
    elem:
    let
      package = elem.package or elem;
      settings = elem.settings or { };
    in
    lib.nameValuePair package.addonId (
      {
        installation_mode = "force_installed";
        install_url = "file://${package.outPath}/share/mozilla/${extensionPath}/${package.addonId}.xpi";
      }
      // settings
    )
  ) cfg.globalExtensions
)
```

Mikilio commented:
@khaneliman I applied your suggestions.

khaneliman commented on modules/programs/firefox/mkFirefoxModule.nix:
```diff
@@ -1132,17 +1197,38 @@ in
+                  }
+                  // settings
+                )
+              ) cfg.globalExtensions
```
This keeps invalid entries from forcing `package.addonId` during policy generation, so the assertion above can report the intended error instead of a raw `attribute addonId missing` eval failure.

```suggestion
              ) (builtins.filter (elem: (elem.package or elem) ? addonId) cfg.globalExtensions)
```

khaneliman commented:
I also tested the happy path and missing-addonId assertion locally. Since this adds a new test file, GitHub cannot apply it as an inline suggestion from the current diff. Patch used below.

Verified with:

```console
nix run path:.#tests -- firefox-global-extensions-assertions firefox-policies floorp-global-extensions-assertions floorp-policies librewolf-global-extensions-assertions librewolf-policies
```

All 6 passed.

```diff
diff --git a/tests/modules/programs/firefox/common.nix b/tests/modules/programs/firefox/common.nix
index 46663547d..31df5f485 100644
--- a/tests/modules/programs/firefox/common.nix
+++ b/tests/modules/programs/firefox/common.nix
@@ -22,6 +22,7 @@ builtins.mapAttrs
     "${name}-deprecated-native-messenger" = ./deprecated-native-messenger.nix;
     "${name}-null-package" = ./null-package.nix;
     "${name}-final-package" = ./final-package.nix;
+    "${name}-global-extensions-assertions" = ./global-extensions-assertions.nix;
     "${name}-policies" = ./policies.nix;
     "${name}-profiles-bookmarks" = ./profiles/bookmarks;
     "${name}-profiles-bookmarks-attrset" = ./profiles/bookmarks/attrset.nix;
diff --git a/tests/modules/programs/firefox/global-extensions-assertions.nix b/tests/modules/programs/firefox/global-extensions-assertions.nix
new file mode 100644
index 000000000..3e8f91201
--- /dev/null
+++ b/tests/modules/programs/firefox/global-extensions-assertions.nix
@@ -0,0 +1,25 @@
+modulePath:
+{ config, lib, ... }:
+
+let
+  firefoxMockOverlay = import ./setup-firefox-mock-overlay.nix modulePath;
+
+  extensionWithoutAddonId = config.lib.test.mkStubPackage {
+    name = "extension-without-addon-id";
+  };
+in
+{
+  imports = [ firefoxMockOverlay ];
+
+  config = lib.mkIf config.test.enableBig (
+    lib.setAttrByPath modulePath {
+      enable = true;
+      globalExtensions = [ extensionWithoutAddonId ];
+    }
+    // {
+      test.asserts.assertions.expected = [
+        "${lib.showOption modulePath}.globalExtensions requires each package to expose addonId in passthru."
+      ];
+    }
+  );
+}
diff --git a/tests/modules/programs/firefox/policies.nix b/tests/modules/programs/firefox/policies.nix
index 23b6a7ad5..8063d354c 100644
--- a/tests/modules/programs/firefox/policies.nix
+++ b/tests/modules/programs/firefox/policies.nix
@@ -11,6 +11,16 @@ let
   firefoxMockOverlay = import ./setup-firefox-mock-overlay.nix modulePath;
 
   darwinPath = "Applications/${cfg.darwinAppName}.app/Contents/Resources";
+
+  uBlockStubPkg = config.lib.test.mkStubPackage {
+    name = "ublock-origin-dummy";
+    extraAttrs.addonId = "uBlock0@raymondhill.net";
+  };
+
+  privacyBadgerStubPkg = config.lib.test.mkStubPackage {
+    name = "privacy-badger-dummy";
+    extraAttrs.addonId = "jid1-MnnxcxisBPnSXQ@jetpack";
+  };
 in
 {
   imports = [ firefoxMockOverlay ];
@@ -22,6 +32,16 @@ in
       policies = {
         BlockAboutConfig = true;
       };
+      globalExtensions = [
+        uBlockStubPkg
+        {
+          package = privacyBadgerStubPkg;
+          settings = {
+            default_area = "menupanel";
+            private_browsing = true;
+          };
+        }
+      ];
       package = pkgs.${cfg.wrappedPackageName}.override {
         extraPolicies = {
           DownloadDirectory = "/foo";
@@ -62,6 +82,31 @@ in
           if [[ $downloadDirectory_actual_value != "\"/foo\"" ]]; then
             fail "Expected '${config_file}' to set 'policies.DownloadDirectory' to \"/foo\""
           fi
+
+          ublock_installation_mode="$($jq --arg extid "${uBlockStubPkg.addonId}" ".policies.ExtensionSettings[\$extid].installation_mode" ${config_file})"
+
+          if [[ $ublock_installation_mode != "\"force_installed\"" ]]; then
+            fail "Expected '${config_file}' to force-install ${uBlockStubPkg.addonId}"
+          fi
+
+          ublock_install_url="$($jq --arg extid "${uBlockStubPkg.addonId}" ".policies.ExtensionSettings[\$extid].install_url" ${config_file})"
+          expected_ublock_install_url="\"file://${uBlockStubPkg}/share/mozilla/extensions/{ec8030f7-c20a-464f-9b0e-13a3a9e97384}/${uBlockStubPkg.addonId}.xpi\""
+
+          if [[ $ublock_install_url != "$expected_ublock_install_url" ]]; then
+            fail "Expected '${config_file}' to set install_url for ${uBlockStubPkg.addonId}"
+          fi
+
+          privacy_badger_private_browsing="$($jq --arg extid "${privacyBadgerStubPkg.addonId}" ".policies.ExtensionSettings[\$extid].private_browsing" ${config_file})"
+
+          if [[ $privacy_badger_private_browsing != "true" ]]; then
+            fail "Expected '${config_file}' to preserve custom settings for ${privacyBadgerStubPkg.addonId}"
+          fi
+
+          privacy_badger_default_area="$($jq --arg extid "${privacyBadgerStubPkg.addonId}" ".policies.ExtensionSettings[\$extid].default_area" ${config_file})"
+
+          if [[ $privacy_badger_default_area != "\"menupanel\"" ]]; then
+            fail "Expected '${config_file}' to preserve default_area for ${privacyBadgerStubPkg.addonId}"
+          fi
         '';
     }
   );

```

Mikilio commented:
I applied your patch, rebased, and credited you @khaneliman

khaneliman reviewed (commented):
Looks good from my side. One small follow-up to handle separately: `globalExtensions` probably wants the same `package != null` handling/assertion that `languagePacks` already has as prior art, since both only take effect through the policy/wrapper path. Not blocking this PR, but worth tightening up in a follow-up.
